### PR TITLE
Synthetic benchmark for Aruco/Charuco/chessboard detection

### PR DESCRIPTION
**Pull request by AleksandrPanov**

### CODE REVIEW

**Comment by asmorkalov** [comment]
General notes:
- Naming is poor. Should be charcoal-thyntetic-benchmark or something meaningful.
- Command line parameters are not used.
- Distortion model as soon as board params should be configured with file, but bot hardcoded.

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
+#!/usr/bin/env python
+
+"""synthetic_benchmark.py
+Usage example:
+python augmentation_benchmark.py -o out.yaml -p path
```
Comment (asmorkalov) [inline python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: Does not match script name.

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+    rvec, tvec = cv.composeRT(rotPitch, np.zeros((3, 1), np.float64),
+                              rotYaw, np.zeros((3, 1), np.float64))[:2]
+
+    rvec, tvec = cv.composeRT(rvecTmp, np.zeros((3, 1), np.float64),
+                              rvec, np.zeros((3, 1), np.float64))[:2]
+    tvec = np.array([[0.], [0.], [distance]])
+    return rvec, tvec
+
+
+def get_coord(num_rows, num_cols, start_x=0, start_y=0):
```
Comment (asmorkalov) [inline python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: what coordinates? please rename.

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+class UndistortFisheyeTransform:
+    def __init__(self, *, img_size):
+        self.cameraMatrix = np.eye(3, 3, dtype=np.float64)
+        self.cameraMatrix[0, 0] = img_size[0]
+        self.cameraMatrix[1, 1] = img_size[0]
+        self.cameraMatrix[0, 2] = img_size[0] / 2
+        self.cameraMatrix[1, 2] = img_size[0] / 2
+        self.distCoeffs = np.zeros((4, 1), np.float64)
+        self.distCoeffs[0] = -0.5012997
+        self.distCoeffs[1] = -0.50116057
```
Comment (asmorkalov) [inline python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: I propose to run the camera parameters from file, e.g. xml generated by calibration.

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+    def detect_and_check_charuco(self, type_dist):
+        pass
+
+
+def main():
+    # parse command line options
+    parser = argparse.ArgumentParser(description="augmentation benchmark", add_help=False)
+    parser.add_argument("-H", "--help", help="show help", action="store_true", dest="show_help")
+    parser.add_argument("-o", "--output", help="output file", default="out.yaml", action="store", dest="output")
+    parser.add_argument("-p", "--path", help="input dataset path", default="", action="store",
```
Comment (asmorkalov) [inline python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: Default value does not work:
```
python3.8 /home/alexander/Projects/OpenCV/opencv_benchmarks/python_benchmarks/synthetic_benchmark/synthetic_benchmark.py
Traceback (most recent call last):
  File "/home/alexander/Projects/OpenCV/opencv_benchmarks/python_benchmarks/synthetic_benchmark/synthetic_benchmark.py", line 488, in <module>
    main()
  File "/home/alexander/Projects/OpenCV/opencv_benchmarks/python_benchmarks/synthetic_benchmark/synthetic_benchmark.py", line 472, in main
    list_folders = next(os.walk(dataset_path))[1]
StopIteration
```

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+        metric = TypeNorm.l2
+    elif args.metric == "intersection_over_union":
+        metric = TypeNorm.intersection_over_union
+
+    cell_img_size = 100
+    board_size = (5, 5)
+    background = BackGroundObject(num_rows=700, num_cols=700)
+    background.show()
+
+    charuco_object = SyntheticCharuco(board_size=board_size, cell_img_size=cell_img_size)
```
Comment (asmorkalov) [inline python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: Board geometry, type and dict should be selected with command line params, or better, with config file and promoted to both detector and generator.

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+                        dest="metric", choices=['l1', 'l2', 'l_inf', 'intersection_over_union'], type=str)
+
+    args = parser.parse_args()
+    show_help = args.show_help
+    if show_help:
+        parser.print_help()
+        return
+    output = args.output
+    dataset_path = args.dataset_path
+    accuracy = args.accuracy
```
Comment (asmorkalov) [inline python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: Not used in the code.

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+            points = points[:, :-1]
+        points = points.transpose()
+        res_points = np.dot(project_mat, points)
+        res_points[0] += self.rot_mat[0, 2]
+        res_points[1] += self.rot_mat[1, 2]
+        res_points = res_points.transpose()
+        return res_points
+
+
+class BluerTransform(TransformObject):
```
Comment (asmorkalov) [inline python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: BlurTransform

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+
+"""synthetic_benchmark.py
+Usage example:
+python augmentation_benchmark.py -p path
+-H, --help - show help
+-o, --output - the path to the output of the dataset or detect statistics
+-p, --path - input dataset path
+-a, --accuracy - input accuracy (default 20 pixels)
+--metric - input norm (default l_inf)
+"""
```
Comment (asmorkalov) [inline python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: Options are out of date.

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py [hunk 9]
```diff
@@ -0,0 +1,488 @@
...
+        res_points[0] += self.rot_mat[0, 2]
+        res_points[1] += self.rot_mat[1, 2]
+        res_points = res_points.transpose()
+        return res_points
+
+
+class BluerTransform(TransformObject):
+    def __init__(self, *, ksize=(5, 5)):
+        self.ksize = ksize
+        self.name = "bluer"
```
Comment (asmorkalov) [inline python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: here too.

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py [hunk 10]
```diff
@@ -0,0 +1,488 @@
...
+        cv.aruco.drawDetectedMarkers(image, aruco)
+        chessboard_corners = self.chessboard_corners.reshape(-1, 1, 2)
+        cv.aruco.drawDetectedCornersCharuco(image, chessboard_corners)
+        cv.imshow("SyntheticCharuco", image)
+        cv.waitKey(wait_key)
+
+    def write(self, path="test", filename="test"):
+        for name in self.fields:
+            self.fields[name] = getattr(self, name)
+        with open(path+"/"+filename+'.txt', 'w') as fp:
```
Comment (asmorkalov) [inline python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: `.json`

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py [hunk 11]
```diff
@@ -0,0 +1,488 @@
...
+            json.dump(self.fields, fp, cls=NumpyEncoder)
```
Comment (asmorkalov) [inline python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: I propose to dump array of boards to be able to generate and check several boards on the same image

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py [hunk 12]
```diff
@@ -0,0 +1,488 @@
...
+        res = np.zeros(6, dtype=np.float64)
+        for config in configs:
+            charuco_object.read(dataset_path+'/'+folder, config.split('/')[-1].split('\\')[-1].split('.')[0])
+            #charuco_object.show()
+            charuco_checker = CharucoChecker(charuco_object)
+            # res = charuco_checker.detect_and_check_aruco()
+            res += np.array(charuco_checker.detect_and_check_charuco())
+        print(folder)
+        print("detected", res[0]/res[1], "total", res[1], "distance", res[2]/max(res[1], 1),
+              "detected", res[3]/res[4], "total", res[4], "distance", res[5]/max(res[4], 1),)
```
Comment (asmorkalov) [inline python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: Let's format it as human-readable table.

**Comment by asmorkalov** [comment]
General notes:
- Command line options are not obvious. I propose to use generate->detect->assess as default strategy.
- I propose to output the final statistics as human readable table + csv for Excel. You can use pandas for it.
- `--metric` does not make sense. I propose output all options in the table / csv.
- Let's add backgroud with random noise and solt / pepper noise or Gaussian noise on top of the image. It makes the images more realistic.

**Comment by asmorkalov** [comment]
```
alexander@paradox:~/tmp/tmp$ python3.8 /mnt/projects/Projects/OpenCV/opencv_benchmarks/python_benchmarks/synthetic_benchmark/synthetic_benchmark.py -o $PWD
Traceback (most recent call last):
  File "/mnt/projects/Projects/OpenCV/opencv_benchmarks/python_benchmarks/synthetic_benchmark/synthetic_benchmark.py", line 616, in <module>
    main()
  File "/mnt/projects/Projects/OpenCV/opencv_benchmarks/python_benchmarks/synthetic_benchmark/synthetic_benchmark.py", line 603, in main
    list_folders = next(os.walk(dataset_path))[1]
StopIteration
```

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py (hunk 11 above)
Comment (asmorkalov) [reply python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: also it should be `.json` file

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py (hunk 10 above)
Comment (AleksandrPanov) [reply python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: fixed

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py (hunk 9 above)
Comment (AleksandrPanov) [reply python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: fixed

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py (hunk 7 above)
Comment (AleksandrPanov) [reply python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: fixed

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py (hunk 11 above)
Comment (AleksandrPanov) [reply python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: added

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py (hunk 6 above)
Comment (AleksandrPanov) [reply python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: fixed

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py (hunk 1 above)
Comment (AleksandrPanov) [reply python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: fixed

File: python_benchmarks/synthetic_benchmark/synthetic_benchmark.py (hunk 5 above)
Comment (AleksandrPanov) [reply python_benchmarks/synthetic_benchmark/synthetic_benchmark.py]: added

**Review by asmorkalov — APPROVED** [review]
:+1:

**Comment by asmorkalov** [comment]
Remaining items:
- Custom image as background.
- CSV output and human-readable results in console.

**Pull request merged**